### PR TITLE
Make BackendSampler set shots and stddev_upper_bound of returned QuasiDistribution

### LINKED ISSUE #9311: QuasiDistribution returned by BackendSampler has incorrect shots
### Environment

- **Qiskit Terra version**: 07e0a2fc79bada7c1fbf0594f4ad33934f70b7e
- **Python version**: 3.10.8
- **Operating system**: Arch Linux

### What is happening?

See title.

### How can we reproduce the issue?

```python
from qiskit.circuit import QuantumCircuit
from qiskit.primitives import BackendSampler
from qiskit.providers.fake_provider import FakeLagos

circuit = QuantumCircuit(2)
circuit.measure_all()

backend = FakeLagos()
sampler = BackendSampler(backend)
result = sampler.run(circuit, shots=100).result()
quasis = result.quasi_dists[0]
print(quasis.shots)
```
```
None
```

### What should happen?

It should print "100", not "None".

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by kevinsung**
### Summary
Fixes #9311

### CODE REVIEW

**Review by Cryoris — COMMENTED** [review]
The shots setting LGTM, but I'm less sure about the `stddev_upper_bound` since 1/N is only the asymptotic behavior, right? As the reference Sampler also doesn't set this I would be happy to merge without the `stddev_upper_bound` -- otherwise maybe someone more familiar with the `QuasiDistribution` class can have a look 🙂

**Comment by kevinsung** [comment]
The logic for `stddev_upper_bound` is the same as at https://github.com/Qiskit/qiskit-ibm-runtime/blob/ce2e94028bfd60ce99c2fa1d5a9d56601e481e34/qiskit_ibm_runtime/utils/sampler_result_decoder.py#L32-L44

Maybe @t-imamichi or @nonhermitian can comment on the correctness. The `stddev_upper_bound` attribute is poorly documented, but my understanding is that it represents an upper bound on the standard deviation of the expected value of the operator `ZZ...Z`.

**Comment by Cryoris** [comment]
Ok, but if it were a bound on the stddev of the `ZZ...` operator then assume `N=2` shots, measuring once +1 and once -1. Then you have a stddev of 1 but the upper bound would be `1/N = 1/2`? That's why I'm a little confused about this argument...

**Comment by kevinsung** [comment]
Consider a quantum state `|ψ⟩` for which the operator `ZZ...Z` is measured N times, yielding a collection of N numbers `[+1, +1, -1, ...]`. The confusion arises because there are two distinct quantities that may both be referred to as the "standard deviation":
1. The standard deviation of the collection of N numbers `[+1, +1, -1, ...]`.
2. Let A be the average of the N numbers `[+1, +1, -1, ...]`. Then A can be considered to be drawn from a probability distribution P representing the result of an experiment in which `|ψ⟩` is prepared and measured N times (repeating this experiment would yield different values for A). The "standard deviation" could refer to the standard deviation of P.

In your comment above you refer to (1), but `stddev_upper_bound` refers to (2).

**Comment by nonhermitian** [comment]
Because the output can be a quasi-distribution with negative values one cannot use standard variance formulas. One can however get an upper-bound on this as described here: https://arxiv.org/abs/2006.14044,
valid only for operators |O| <= 1

**Comment by kevinsung** [comment]
> Because the output can be a quasi-distribution with negative values one cannot use standard variance formulas. One can however get an upper-bound on this as described here: https://arxiv.org/abs/2006.14044, valid only for operators |O| <= 1

In this case it's a probability distribution though, since it's calculated as
```python
prob_dist = {k: v / shots for k, v in count.int_outcomes().items()}
```

**Comment by mtreinish** [comment]
> 
> In this case it's a probability distribution though, since it's calculated as
> 
> https://github.com/Qiskit/qiskit-terra/blob/b271adb3087205f05cbc11288e582dfd20a48fa7/qiskit/primitives/backend_sampler.py#L169

Yeah for the `BackendSampler` it's returning a normal probability distribution because there is no error mitigation being done as part of the generic implementation for any backend it does. It just computes the probability via division, we returned a `QuasiDistribution` because that's what the abstract sampler api dictates https://github.com/Qiskit/qiskit-terra/blob/c2215ae87b516e57357075d69753489f056cbabc/qiskit/primitives/base/sampler_result.py#L44

File: releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
```
Comment (ikkoham) [inline releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml:2]:
```suggestion
features:
```

File: releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+  - |
+    Fixed a bug where :class:`~.BackendSampler` failed to set the ``shots``
```
Comment (ikkoham) [inline releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml:4]:
```suggestion
    :class:`~.BackendSampler` set the ``shots``
```

**Review by ikkoham — APPROVED** [review]
sorry I forgot reviewing. LGTM. In this case, gamma is always 1, so this definition of the upper bound of standard deviation is correct.

Just one comment: this is not a bug, but enhancement.

File: releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml (hunk 2 above)
Comment (kevinsung) [reply releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml]: Done.

File: releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml (hunk 1 above)
Comment (kevinsung) [reply releasenotes/notes/backend-sampler-shots-c233d5a3965e0c11.yaml]: Done.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**